scanner/fetcher.go

```diff
@@ -0,0 +1,191 @@
+	sth *ct.SignedTreeHead
+
+	// TODO(pavelkalinnikov): Consider log.Logger instead.
+	Log func(msg string)
+}
+
+// EntryBatch represents a contiguous range of entries of the Log.
+type EntryBatch struct {
+	start   int64          // LeafIndex of the first entry in the range.
+	entries []ct.LeafEntry // Entries of the range.
+}
+
+// fetchRange represents a range of certs to fetch from a CT log.
+type fetchRange struct {
+	start int64 // inclusive
+	end   int64 // inclusive
+}
+
+// NewFetcher creates a Fetcher instance using client to talk to the log,
+// taking configuration options from opts.
+func NewFetcher(cli *client.LogClient, opts *FetcherOptions) *Fetcher {
+	fetcher := &Fetcher{cli: cli, opts: opts}
+	if opts.Quiet {
+		fetcher.Log = func(msg string) {}
+	} else {
+		fetcher.Log = func(msg string) { log.Print(msg) }
+	}
+	return fetcher
+}
+
+// Prepare caches the latest Log's STH in the Fetcher and returns it. It also
+// adjusts the entry range to fit the size of the tree.
+func (f *Fetcher) Prepare(ctx context.Context) (*ct.SignedTreeHead, error) {
+	sth, err := f.cli.GetSTH(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("GetSTH() failed: %v", err)
+	}
+	f.Log(fmt.Sprintf("Got STH with %d certs", sth.TreeSize))
+	if f.opts.EndIndex == 0 || f.opts.EndIndex > int64(sth.TreeSize) {
+		f.opts.EndIndex = int64(sth.TreeSize)
+	}
+	f.sth = sth
+	return sth, nil
+}
+
+// Run performs fetching of the Log. Blocks until scanning is complete or
+// context is cancelled. For each successfully fetched batch, pushes it to the
+// out channel.
+func (f *Fetcher) Run(ctx context.Context, out chan<- EntryBatch) error {
+	f.Log("Starting up Fetcher...\n")
+
+	if f.sth == nil {
+		if _, err := f.Prepare(ctx); err != nil {
+			return err
+		}
+	}
+
+	jobs := f.genJobs(ctx)
+
+	// Run fetcher workers.
+	var wg sync.WaitGroup
+	for w, cnt := 0, f.opts.ParallelFetch; w < cnt; w++ {
+		wg.Add(1)
+		go func(idx int) {
+			defer wg.Done()
+			f.runWorker(ctx, jobs, out)
+			f.Log(fmt.Sprintf("Fetcher worker %d finished", idx))
+		}(w)
+	}
+	wg.Wait()
+
+	return nil
+}
+
+// getJobs returns a channel of fetching jobs, each job is a Log entry range.
+// Can be stopped using ctx.
+func (f *Fetcher) genJobs(ctx context.Context) <-chan fetchRange {
+	start, end := f.opts.StartIndex, f.opts.EndIndex
+	batch := int64(f.opts.BatchSize)
+
+	jobs := make(chan fetchRange)
+	go func() {
+		defer close(jobs)
+		for start < end {
+			batchEnd := min(start+batch, end)
+			next := fetchRange{start, batchEnd - 1}
+			select {
+			case <-ctx.Done():
+				f.Log(fmt.Sprintf("genJobs cancelled: %v", ctx.Err()))
+				return
+			case jobs <- next:
+			}
+			start = batchEnd
+		}
+	}()
+	return jobs
+}
+
+// Worker function for fetcher jobs.
+// Accepts cert ranges to fetch over the jobs channel, and if the fetch is
+// successful sends the []entryInfo slice to the batches channel. Will retry
+// failed attempts to retrieve ranges indefinitely.
```

Comment:
/indefinitely/until the context is cancelled/?

Reply:
Done.